File: core/src/main/java/mifos/org/mobilewallet/core/domain/usecase/FetchClientData.java
package mifos.org.mobilewallet.core.domain.usecase;

import javax.inject.Inject;

import mifos.org.mobilewallet.core.base.UseCase;
import mifos.org.mobilewallet.core.data.fineract.repository.FineractRepository;
import mifos.org.mobilewallet.core.domain.model.ClientDetails;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by naman on 17/6/17.
 */

public class FetchClientData extends UseCase<FetchClientData.RequestValues,
        FetchClientData.ResponseValue> {

    private final FineractRepository fineractRepository;

    @Inject
    public FetchClientData(FineractRepository fineractRepository) {
        this.fineractRepository = fineractRepository;
    }


    @Override
    protected void executeUseCase(RequestValues requestValues) {

        if (requestValues != null) {
            fineractRepository.getClientDetails(requestValues.clientid)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(Schedulers.io())
                    .subscribe(new Subscriber<ClientDetails>() {
                        @Override
                        public void onCompleted() {

                        }

                        @Override
                        public void onError(Throwable e) {
                            getUseCaseCallback().onError("Error fetching client data");
                        }

                        @Override
                        public void onNext(ClientDetails clientDetails) {
                            getUseCaseCallback().onSuccess(new ResponseValue(clientDetails));
                        }
                    });
        } else {
            fineractRepository.getClientDetails()
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(Schedulers.io())
                    .subscribe(new Subscriber<ClientDetails>() {
                        @Override
                        public void onCompleted() {

                        }

                        @Override
                        public void onError(Throwable e) {
                            getUseCaseCallback().onError("Error fetching client data");
                        }

                        @Override
                        public void onNext(ClientDetails clientDetails) {
                            if (clientDetails != null) {
                                getUseCaseCallback().onSuccess(new ResponseValue(clientDetails));
                            } else {
                                getUseCaseCallback().onError("No client found");
                            }
                        }
                    });
        }
    }



    public static final class RequestValues implements UseCase.RequestValues {

        private final long clientid;

        public RequestValues(long clientid) {
            this.clientid = clientid;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {

        private final ClientDetails user;

        public ResponseValue(ClientDetails user) {
            this.user = user;
        }

        public ClientDetails getUserDetails() {
            return user;
        }
    }
}

File: core/data/src/main/java/org/mifospay/core/data/domain/usecase/client/FetchClientData.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/mobile-wallet/blob/master/LICENSE.md
 */
package org.mifospay.core.data.domain.usecase.client

import com.mifospay.core.model.entity.Page
import com.mifospay.core.model.entity.client.Client
import org.mifospay.core.data.base.UseCase
import org.mifospay.core.data.fineract.entity.mapper.ClientDetailsMapper
import org.mifospay.core.data.fineract.repository.FineractRepository
import org.mifospay.core.data.util.Constants
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

class FetchClientData(
    private val fineractRepository: FineractRepository,
    private val clientDetailsMapper: ClientDetailsMapper,
) : UseCase<FetchClientData.RequestValues, FetchClientData.ResponseValue>() {

    override fun executeUseCase(requestValues: RequestValues) {
        requestValues.clientId?.let { clientId ->
            fineractRepository.getSelfClientDetails(clientId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(
                    object : Subscriber<Client>() {
                        override fun onCompleted() {}
                        override fun onError(e: Throwable) {
                            useCaseCallback.onError(Constants.ERROR_FETCHING_CLIENT_DATA)
                        }

                        override fun onNext(client: Client) {
                            useCaseCallback.onSuccess(
                                ResponseValue(clientDetailsMapper.transform(client)),
                            )
                        }
                    },
                )
        } ?: run {
            fineractRepository.selfClientDetails
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(
                    object : Subscriber<Page<Client>>() {
                        override fun onCompleted() {}
                        override fun onError(e: Throwable) {
                            useCaseCallback.onError(Constants.ERROR_FETCHING_CLIENT_DATA)
                        }

                        override fun onNext(client: Page<Client>) {
                            if (client.pageItems.size != 0) {
                                useCaseCallback.onSuccess(
                                    ResponseValue(clientDetailsMapper.transform(client.pageItems[0])),
                                )
                            } else {
                                useCaseCallback.onError(Constants.NO_CLIENT_FOUND)
                            }
                        }
                    },
                )
        }
    }

    data class RequestValues(val clientId: Long?) : UseCase.RequestValues
    data class ResponseValue(
        val clientDetails: com.mifospay.core.model.domain.client.Client,
    ) : UseCase.ResponseValue
}
